File: Striver/DP/Q15.py
# Partition equal subset sum

# recursive solution
def subset_sum(arr, target, n):
    if target == 0:
        return True
    if n == 0:
        return arr[0] == target
    not_pick = subset_sum(arr, target, n-1)
    pick = False
    if target >= arr[n]:
        pick = subset_sum(arr, target-arr[n], n-1)
    return pick or not_pick

def canPartition(nums):
    total = sum(nums)
    if total % 2 != 0:
        return False
    return subset_sum(nums, total//2, len(nums)-1)

# memoization
def subset_sum_memo(arr, target, n, dp):
    if target == 0:
        return True
    if n == 0:
        return arr[0] == target
    if dp[n][target] != -1:
        return dp[n][target]
    not_pick = subset_sum_memo(arr, target, n-1, dp)
    pick = False
    if target >= arr[n]:
        pick = subset_sum_memo(arr, target-arr[n], n-1, dp)
    dp[n][target] = pick or not_pick
    return dp[n][target]

def canPartition_memo(nums):
    total = sum(nums)
    if total % 2 != 0:
        return False
    dp = [[-1 for i in range(total//2+1)] for j in range(len(nums)+1)]
    return subset_sum_memo(nums, total//2, len(nums)-1, dp)

# tabulation
def subset_sum_tab(arr, target, n):
    dp = [[False for i in range(target+1)] for j in range(n+1)]
    for i in range(n+1):
        dp[i][0] = True
    dp[0][arr[0]] = True
    for i in range(1, n+1):
        for j in range(1, target+1):
            not_pick = dp[i-1][j]
            pick = False
            if j >= arr[i]:
                pick = dp[i-1][j-arr[i]]
            dp[i][j] = pick or not_pick
    return dp[n][target]

def canPartition_tab(nums):
    total = sum(nums)
    if total % 2 != 0:
        return False
    return subset_sum_tab(nums, total//2, len(nums)-1)

# space optimization
def subset_sum_tab_opt(arr, target, n):
    dp = [False for i in range(target+1)]
    curr = [False for i in range(target+1)]
    dp[0] = True
    curr[0] = True
    dp[arr[0]] = True
    for i in range(1, n+1):
        for j in range(target, arr[i]-1, -1):
            not_pick = dp[j]
            pick = False
            if j >= arr[i]:
                pick = dp[j-arr[i]]
            curr[j] = pick or not_pick
        dp = curr
    return dp[target]

def canPartition_tab_opt(nums):
    total = sum(nums)
    if total % 2 != 0:
        return False
    return subset_sum_tab_opt(nums, total//2, len(nums)-1)

arr = [1, 5, 11, 5]
print(canPartition(arr))
print(canPartition_memo(arr))
print(canPartition_tab(arr))
print(canPartition_tab_opt(arr))

File: Striver/DP/Q15.java
package Striver.DP;

public class Q15{
    // Partition equal subset sum

    // recursive solution - Time: O(2^n) Space: O(n)
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for(int i: nums) sum += i;
        if(sum%2 != 0) return false;
        return subsetSum(nums, sum/2, 0);
    }

    public boolean subsetSum(int[] arr, int target, int n){
        if(target == 0) return true;
        if(n == 0) return arr[0] == target;
        boolean notPick = subsetSum(arr, target, n-1);
        boolean pick = false;
        if(target >= arr[n]) pick = subsetSum(arr, target-arr[n], n-1);
        return notPick || pick;
    }

    // memoization - Time: O(n*target) Space: O(n*target)
    public boolean canPartitionMemo(int[] nums) {
        int sum = 0;
        for(int i: nums) sum += i;
        if(sum%2 != 0) return false;
        Boolean[][] dp = new Boolean[nums.length][sum/2+1];
        return subsetSumMemo(nums, sum/2, nums.length-1, dp);
    }

    public boolean subsetSumMemo(int[] arr, int target, int n, Boolean[][] dp){
        if(target == 0) return true;
        if(n == 0) return arr[0] == target;
        if(dp[n][target] != null) return dp[n][target];
        boolean notPick = subsetSumMemo(arr, target, n-1, dp);
        boolean pick = false;
        if(target >= arr[n]) pick = subsetSumMemo(arr, target-arr[n], n-1, dp);
        return dp[n][target] = notPick || pick;
    }

    // tabulation - Time: O(n*target) Space: O(n*target)
    public boolean canPartitionTab(int[] nums) {
        int sum = 0;
        for(int i: nums) sum += i;
        if(sum%2 != 0) return false;
        boolean[][] dp = new boolean[nums.length+1][sum/2+1];
        for(int i=0; i<=nums.length; i++) dp[i][0] = true;
        dp[0][nums[0]] = true;
        for(int i=1; i<=nums.length; i++) {
            for(int j=1; j<=sum/2; j++) {
                boolean notPick = dp[i-1][j];
                boolean pick = false;
                if(j >= nums[i-1]) pick = dp[i-1][j-nums[i-1]];
                dp[i][j] = notPick || pick;
            }
        }
        return dp[nums.length][sum/2];
    }

    // space optimization - Time: O(n*target) Space: O(target)
    public boolean canPartitionSpaceOpt(int[] arr) {
        int sum = 0;
        for(int i: arr) sum += i;
        if(sum%2 != 0) return false;
        int target = sum/2;
        int n = arr.length;
        boolean[] dp = new boolean[target+1];
        boolean[] curr = new boolean[target+1];
        dp[0] = true;
        curr[0] = true;
        dp[arr[0]] = true;
        for(int i=1; i<=n; i++) {
            for(int j=1; j <= target; j++) {
                boolean notPick = dp[j];
                boolean pick = false;
                if(j >= arr[i-1]) pick = dp[j-arr[i-1]];
                curr[j] = notPick || pick;
            }
            dp = curr;
        }
        return dp[target];
    }
}